code/modules/power/smes.dm

```diff
@@ -206,156 +206,83 @@
-</div>
-"}
-
-	user.Browse(t, "window=smes;size=400x340")
-	onclose(user, "smes")
-	return
-
-/obj/machinery/power/smes/Topic(href, href_list)
-	..()
-
-	if (usr.stat || usr.restrained() )
+/obj/machinery/power/smes/ui_state(mob/user)
+	return tgui_default_state
+
+/obj/machinery/power/smes/ui_status(mob/user, datum/ui_state/state)
+	return min(
+		state.can_use_topic(src, user),
+		tgui_broken_state.can_use_topic(src, user),
+		tgui_not_incapacitated_state.can_use_topic(src, user)
+	)
+
+/obj/machinery/power/smes/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Smes", src.name)
+		ui.open()
+
+/obj/machinery/power/smes/ui_data(mob/user)
+	var/list/data = list()
+	data["capacityPercent"] = (src.charge / src.capacity * 100)
+	data["charge"] = src.charge
+	data["formattedCharge"] = engineering_notation(src.charge)
+	data["inputAttempt"] = src.chargemode
+	data["inputting"] = src.charging
+	data["inputStartUp"] = src.chargecount //smes waits before charging
+	data["inputLevel"] = src.chargelevel
+	data["inputLevelMax"] = SMESMAXCHARGELEVEL
+	data["inputAvailable"] = src.lastexcess
+	data["outputAttempt"] = src.online
+	data["outputting"] = src.loaddemand
+	data["outputLevel"] = src.output
+	data["outputLevelMax"] = SMESMAXOUTPUT
+	return data
+
+/obj/machinery/power/smes/ui_act(action, params)
+	if(..())
 		return
-
-	if (( usr.using_dialog_of(src) && ((get_dist(src, usr) <= 1) && istype(src.loc, /turf))) || (isAI(usr) || issilicon(usr)))
-		if (href_list["close"])
-			usr.Browse(null, "window=smes")
-			src.remove_dialog(usr)
-			return
-
-		else if ( href_list["cmode"] )
-			chargemode = !chargemode
+	switch(action)
+		if("tryinput") //toggle charging
+			src.chargemode = !src.chargemode
 			if (!chargemode)
 				charging = 0
-			updateicon()
-
-		else if ( href_list["online"] )
-			online = !online
-			updateicon()
-		else if (href_list["input"])
-			switch (href_list["input"])
-				if ("min")
-					chargelevel = 0
-				if ("max")
-					chargelevel = SMESMAXCHARGELEVEL
-				if ("set")
-					var/newnum = input(usr, "New target charge level? 0 to [SMESMAXCHARGELEVEL].", "SMES Config", chargelevel) as null|num
-					if (newnum)
-						chargelevel	= newnum
-
-			chargelevel = max(0, min(SMESMAXCHARGELEVEL, chargelevel))	// clamp to range
-
-		else if (href_list["output"])
-			switch (href_list["output"])
-				if ("min")
-					output = 0
-				if ("max")
-					output = SMESMAXOUTPUT
-				if ("set")
-					var/newnum = input(usr, "New output level? 0 to [SMESMAXOUTPUT].", "SMES Config", output) as null|num
-					if (newnum)
-						output = newnum
-
-			output = max(0, min(SMESMAXOUTPUT, output))	// clamp to range
-
-
-		src.updateUsrDialog()
-
-	else
-		usr.Browse(null, "window=smes")
-		src.remove_dialog(usr)
-
-	return
+			src.updateicon()
+			. = TRUE
+		if("tryoutput") //toggle output
+			src.online = !src.online
+			src.updateicon()
+			. = TRUE
+		if("input")
```

Comment:
Prefer `verb-noun` for action name, e.g. `set-input`. Not required by our coding style, but good practice for naming actions in general.